snuba/datasets/plans/query_plan.py

```diff
@@ -0,0 +1,68 @@
+from __future__ import annotations
+
+from abc import ABC
+from dataclasses import dataclass
+from typing import Callable, Sequence
+
+from snuba.query import RawQueryResult
+from snuba.query.query_processor import QueryProcessor
+from snuba.request import Request
+
+
+SingleQueryRunner = Callable[[Request], RawQueryResult]
+
+
+@dataclass(frozen=True)
+class StorageQueryPlan:
+    """
+    Provides the directions to execute the query against one storage
+    or multiple joined ones.
+    This is produced by StorageQueryPlanBuilder (provided by the dataset)
+    after the dataset query processing has been performed and the storage
+    has been selected.
+    It embeds the sequence of storage specific QueryProcessors to apply
+    to the query after the the storage has been selected.
+    It also provides a plan execution strategy, in case the query is not
+    one individual query statement (like for split queries).
+    """
+
+    # TODO: When we will have a separate Query class for Snuba Query and
+    # Storage QUery, this plan will also provide the Storage Query. Right
+    # now the storage query is the same mutable object referenced by Request
+    # so no need to add an additional reference here (it would make the query
+    # execution code more confusing).
+    query_processors: Sequence[QueryProcessor]
+    execution_strategy: QueryPlanExecutionStrategy
+
+
+class QueryPlanExecutionStrategy(ABC):
+    """
+    Orchestrate the executions of a query. An example use case is split
+    queries, when the split is done at storage level.
+    It does not know how to run a query against a DB, but it knows how
+    and whether to break down a query and how to assemble the results back.
+    It receives a runner, that takes care of actually executing one statement
+    against the DB.
+    Potentially this could be agnostic to the DB.
+    """
+
+    def execute(self, request: Request, runner: SingleQueryRunner) -> RawQueryResult:
```

Comment:
I would rather pass Query and Settings instead of the Request object (as we do for query processors).
I originally built this method to depend on Query and ReqyestSettings, but, when I introduced this in web/query.py in the query processing logic that became extremely confusing.
The Runner needs the Request object because the accesses elements like the referrer, but worse, it passes it through to other functions that pick up the query from the Request.
So in order to pass Query and RequestSettings to this method I would have had to also pass the Request to the runner (partially applied function), thus all functions down the stack from here would have had access to two query objects. One accessible from the Request object and one passed directly to this method and to runner. They happen to be references to the same object but this is confusing at best and this assumption will miserably break when we break apart SnubaQuery AST from ClickhouseQuery AST.

Passing the Request object around forces us, when we will split Snuba Query from Clickhouse Query to take care of this problem which we could easily forget about otherwise.